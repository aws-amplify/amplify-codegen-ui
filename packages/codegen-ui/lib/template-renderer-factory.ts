/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import { FrameworkOutputManager } from './framework-output-manager';
import { StudioTemplateRenderer } from './studio-template-renderer';

import { RenderTextComponentResponse } from './render-component-response';

/**
 * This class is used to wrap the created of renderers due to each renderer
 * only being used for one component.
 */
export class StudioTemplateRendererFactory<
  TSource,
  TStudioType,
  TOutputManager extends FrameworkOutputManager<TSource>,
  TRenderOutput extends RenderTextComponentResponse,
  TRenderer extends StudioTemplateRenderer<TSource, TStudioType, TOutputManager, TRenderOutput>,
> {
  constructor(private renderer: (component: TStudioType) => TRenderer) {}

  buildRenderer(component: TStudioType): TRenderer {
    return this.renderer(component);
  }
}
