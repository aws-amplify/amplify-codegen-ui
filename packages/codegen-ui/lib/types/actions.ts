/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import { StudioComponentProperty, StateStudioComponentProperty } from './properties';

export type StateReference = StateStudioComponentProperty | MutationActionSetStateParameter;

export type MutationAction = {
  action: 'Amplify.Mutation';
  parameters: {
    state: MutationActionSetStateParameter;
  };
};

export type MutationActionSetStateParameter = {
  componentName: string;
  property: string;
  set: StudioComponentProperty;
};

export type ActionStudioComponentEvent =
  | NavigationAction
  | AuthSignOutAction
  | DataStoreCreateItemAction
  | DataStoreUpdateItemAction
  | DataStoreDeleteItemAction
  | MutationAction;

export type NavigationAction = {
  action: 'Amplify.Navigation';
  parameters: {
    type: StudioComponentProperty;
    url?: StudioComponentProperty;
    anchor?: StudioComponentProperty;
    target?: StudioComponentProperty;
  };
};

export type AuthSignOutAction = {
  action: 'Amplify.AuthSignOut';
  parameters: {
    global: StudioComponentProperty;
  };
};

export type DataStoreCreateItemAction = {
  action: 'Amplify.DataStoreCreateItemAction';
  parameters: {
    model: string;
    fields: {
      [propertyName: string]: StudioComponentProperty;
    };
  };
};

export type DataStoreUpdateItemAction = {
  action: 'Amplify.DataStoreUpdateItemAction';
  parameters: {
    model: string;
    id: StudioComponentProperty;
    fields: {
      [propertyName: string]: StudioComponentProperty;
    };
  };
};

export type DataStoreDeleteItemAction = {
  action: 'Amplify.DataStoreDeleteItemAction';
  parameters: {
    model: string;
    id: StudioComponentProperty;
  };
};
