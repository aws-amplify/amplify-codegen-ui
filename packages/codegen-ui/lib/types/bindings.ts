/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
export type StudioComponentPropertyBinding =
  | StudioComponentDataPropertyBinding
  | StudioComponentStoragePropertyBinding
  | StudioComponentSimplePropertyBinding
  | StudioComponentEventPropertyBinding
  | StudioComponentSlotBinding;

/**
 * These are the primitive value types
 */
export enum StudioComponentPropertyType {
  String = 'String',
  Number = 'Number',
  Boolean = 'Boolean',
  Date = 'Date',
}

export type StudioComponentSimplePropertyBinding = {
  /**
   *  This is the data type for the value that is bound to this property. The default
   * inferred type is string so this will only need to be set if it is not a string
   */
  type: keyof typeof StudioComponentPropertyType;

  defaultValue?: string | undefined;
};

/**
 * This represent the configuration for binding a component property
 * to Amplify specific information
 */
export type StudioComponentDataPropertyBinding = {
  /**
   * This declares where the data is coming from to bind to
   */
  type: 'Data';

  /**
   * This is the value of the data binding
   */
  bindingProperties: StudioComponentDataBindingProperty;
};

/**
 * This represent the configuration for binding a component property
 * to Amplify specific information
 */
export type StudioComponentEventPropertyBinding = {
  /**
   * This declares that the type is of a workflow binding
   */
  type: 'Event';
};

/**
 * This represent the configuration for binding a component property
 * to Amplify specific information
 */
export type StudioComponentStoragePropertyBinding = {
  /**
   * This declares where the data is coming from to bind to
   */
  type: 'Storage';

  /**
   * This is the value of the data binding
   */
  bindingProperties: StudioComponentStorageBindingProperty;
};

/**
 * This represent the configuration for a binding to be a user-defined JSX Element
 */
export type StudioComponentSlotBinding = {
  /**
   * This declares that the binding is a Slot type
   */
  type: 'Amplify.Slot';
};

/**
 * This represents the model and field you want to bind
 * a component property to
 */
export type StudioComponentDataBindingProperty = {
  model: string;
  field?: string;
  predicate?: StudioComponentPredicate;
};

export type StudioComponentPredicate = {
  and?: StudioComponentPredicate[];
  or?: StudioComponentPredicate[];
  not?: StudioComponentPredicate[];
  field?: string;
  operand?: string;
  operator?: string;
  operandType?: string;
};

/**
 * This represents the bucket and key you want to bind a component
 * property to
 */
export type StudioComponentStorageBindingProperty = {
  key: string;
  level?: string;
};

/**
 * These are the types of data binding
 */
export enum StudioComponentPropertyBindingType {
  Data = 'Data',
  Storage = 'Storage',
  Event = 'Event',
}

/**
 * This represent the configuration for binding a component property
 * to Amplify specific information
 */
export type StudioComponentDataConfiguration = {
  model: string;

  sort?: StudioComponentSort[];

  predicate?: StudioComponentPredicate;

  /**
   * This is a collection of Id's that will be always queried from DataStore.
   * This would be used in liu of a predicate.
   */
  identifiers?: string[];
};

export type StudioComponentSort = {
  field: string;
  direction: 'ASC' | 'DESC';
};
