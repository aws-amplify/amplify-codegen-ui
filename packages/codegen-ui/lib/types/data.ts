/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */

export type SchemaEnums = Record<string, SchemaEnum>;
export type SchemaEnum = {
  name: string;
  values: string[];
};

type FieldType = string | { model: string } | { nonModel: string } | { enum: string };

export type DataStoreModelField = {
  name: string;
  type: FieldType;
  isReadOnly: boolean;
  isRequired: boolean;
  isArray: boolean;
};

export type CommonRelationshipType = {
  relatedModelName: string;
};

export type HasManyRelationshipType = {
  type: 'HAS_MANY';
  relatedModelFields: string[];
  canUnlinkAssociatedModel: boolean;
  relatedJoinFieldName?: string;
  relatedJoinTableName?: string;
  belongsToFieldOnRelatedModel?: string;
} & CommonRelationshipType;

export type HasOneRelationshipType = {
  type: 'HAS_ONE';
  associatedFields?: string[];
  isHasManyIndex?: boolean;
} & CommonRelationshipType;

export type BelongsToRelationshipType = {
  type: 'BELONGS_TO';
  associatedFields?: string[];
  isHasManyIndex?: boolean;
} & CommonRelationshipType;

export type GenericDataRelationshipType = HasManyRelationshipType | HasOneRelationshipType | BelongsToRelationshipType;

export type DataFieldDataType =
  | 'ID'
  | 'String'
  | 'Int'
  | 'Float'
  | 'AWSDate'
  | 'AWSTime'
  | 'AWSDateTime'
  | 'AWSTimestamp'
  | 'AWSEmail'
  | 'AWSURL'
  | 'AWSIPAddress'
  | 'Boolean'
  | 'AWSJSON'
  | 'AWSPhone'
  | { enum: string }
  | { model: string }
  | { nonModel: string };

export type GenericDataField = {
  dataType: DataFieldDataType;

  required: boolean;

  readOnly: boolean;

  isArray: boolean;

  relationship?: GenericDataRelationshipType;
};

export type GenericDataModel = {
  fields: { [fieldName: string]: GenericDataField };
  isJoinTable?: boolean;
  primaryKeys: string[];
};

export type GenericDataNonModel = {
  fields: { [fieldName: string]: GenericDataField };
};

export type GenericDataSchema = {
  dataSourceType: 'DataStore';

  models: { [modelName: string]: GenericDataModel };

  enums: { [enumName: string]: { values: string[] } };

  nonModels: { [nonModelName: string]: GenericDataNonModel };
};
