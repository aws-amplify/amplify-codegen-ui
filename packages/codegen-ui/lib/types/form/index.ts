/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */

import { LabelDecorator, StudioFormStyle } from './style';
import { StudioFormFields, StudioFormFieldConfig, StudioGenericFieldConfig } from './fields';
import { GenericSectionalElementConfig, SectionalElementConfig, SectionalElementFields } from './sectional-element';
import { FormDefinition, ModelFieldsConfigs, FieldTypeMapKeys, ButtonConfig } from './form-definition';
import { StudioFieldInputConfig, StudioFormValueMappings, StudioFormInputFieldProperty } from './input-config';
import { StudioFieldPosition } from './position';
import { StudioFormCTA } from './form-cta';
import {
  FormMetadata,
  FieldConfigMetadata,
  StudioFormActionType,
  StudioFormDataType,
  StudioDataSourceType,
} from './form-metadata';

/**
 * This is the base type for all StudioForms
 */
export type StudioForm = {
  id?: string;

  name: string;

  formActionType: StudioFormActionType;

  dataType: StudioFormDataType;

  fields: StudioFormFields;

  sectionalElements: SectionalElementFields;

  style: StudioFormStyle;

  cta: StudioFormCTA;

  labelDecorator?: LabelDecorator;
};

export type FormInputType =
  | 'TextField'
  | 'TextAreaField'
  | 'PasswordField'
  | 'SliderField'
  | 'StepperField'
  | 'SwitchField'
  | 'ToggleButton'
  | 'CheckboxField'
  | 'RadioGroupField'
  | 'PhoneNumberField'
  | 'SelectField'
  | 'NumberField'
  | 'DateField'
  | 'TimeField'
  | 'DateTimeField'
  | 'IPAddressField'
  | 'URLField'
  | 'EmailField'
  | 'JSONField'
  | 'Autocomplete'
  | 'StorageField';

export * from './form-definition-element';
export * from './style';
export * from './form-validation';
export * from './form-cta';

export type {
  GenericSectionalElementConfig,
  SectionalElementConfig,
  SectionalElementFields,
  StudioFormFieldConfig,
  StudioFormActionType,
  StudioDataSourceType,
  FormDefinition,
  FormMetadata,
  FieldConfigMetadata,
  StudioFieldInputConfig,
  StudioFormInputFieldProperty,
  StudioGenericFieldConfig,
  StudioFormFields,
  ModelFieldsConfigs,
  StudioFieldPosition,
  FieldTypeMapKeys,
  StudioFormValueMappings,
  ButtonConfig,
};
