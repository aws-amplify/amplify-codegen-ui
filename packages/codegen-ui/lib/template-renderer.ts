/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import fs from 'fs';
import { FrameworkOutputManager } from './framework-output-manager';
import { FrontendManagerTemplateRenderer } from './frontend-manager-template-renderer';
import { FrontendManagerTemplateRendererFactory } from './template-renderer-factory';

import { FrameworkOutputConfig } from './framework-output-config';
import { RenderTextComponentResponse } from './render-component-response';
import { handleCodegenErrors, InvalidInputError } from './errors';

/**
 * This is a class for genercially rendering FrontendManager templates.
 * The output is determined by the renderer passed into the constructor.
 */
export class FrontendManagerTemplateRendererManager<
  TSource,
  TFrontendManagerType,
  TOutputManager extends FrameworkOutputManager<TSource>,
  TRenderOutput extends RenderTextComponentResponse,
  TRenderer extends FrontendManagerTemplateRenderer<TSource, TFrontendManagerType, TOutputManager, TRenderOutput>,
> {
  constructor(
    private renderer: FrontendManagerTemplateRendererFactory<
      TSource,
      TFrontendManagerType,
      TOutputManager,
      TRenderOutput,
      TRenderer
    >,
    private outputConfig: FrameworkOutputConfig,
  ) {
    const renderPath = this.outputConfig.outputPathDir;
    if (!fs.existsSync(renderPath)) {
      fs.mkdirSync(renderPath);
    }
  }

  @handleCodegenErrors
  renderSchemaToTemplate(component: TFrontendManagerType | undefined): TRenderOutput {
    if (!component) {
      throw new InvalidInputError('Please ensure you have passed in a valid component schema');
    }
    const componentRenderer = this.renderer.buildRenderer(component);
    const result = componentRenderer.renderComponent();
    result.renderComponentToFilesystem(this.outputConfig.outputPathDir);
    return result;
  }

  @handleCodegenErrors
  renderSchemaToTemplates(jsonSchema: TFrontendManagerType[] | undefined) {
    if (!jsonSchema) {
      throw new InvalidInputError('Please ensure you have passed in a valid schema');
    }

    for (const component of jsonSchema) {
      this.renderer.buildRenderer(component).renderComponent();
    }
  }
}
