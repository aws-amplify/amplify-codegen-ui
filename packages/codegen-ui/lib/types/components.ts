/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import { StudioComponentDataConfiguration, StudioComponentPropertyBinding } from './bindings';
import { StudioComponentEvents } from './events';
import { FigmaMetadata } from './figma';
import { StudioComponentProperties } from './properties';

/**
 * This is the base type for all Studio components
 */
export type StudioComponent = {
  /**
   * The name of the customized component
   */
  name?: string;

  /**
   *  This is the unique global identifier for each component
   */
  id?: string;

  /**
   * The id if the component in its source system (Figma, Sketch, etc.)
   */
  sourceId?: string;

  /**
   * This should map to the components available including Amplify
   * UI components and other custom components
   */
  componentType: string;

  /**
   * These are the customized properties
   */
  properties: StudioComponentProperties;

  /**
   * These are the nested components in a composite
   */
  children?: StudioComponentChild[];

  /**
   * The  metatdata gerated by Figma
   */
  figmaMetadata?: FigmaMetadata;

  /**
   * Variants in terms of styles
   */
  variants?: StudioComponentVariant[];

  /**
   * Overrides for primitives
   */
  overrides?: StudioComponentOverrides;

  bindingProperties: {
    [propertyName: string]: StudioComponentPropertyBinding;
  };
  /**
   * These are the collection properties
   */
  collectionProperties?: {
    [propertyName: string]: StudioComponentDataConfiguration;
  };

  events?: StudioComponentEvents;
};

/**
 * This is the child type for Studio components
 */
export type StudioComponentChild = {
  /**
   * This should map to the components available including Amplify
   * UI components and other custom components
   */
  componentType: string;

  /**
   * The unique name of the child element.
   */
  name: string;

  /**
   * These are the customized properties
   */
  properties: StudioComponentProperties;

  /**
   * These are the nested components in a composite
   */
  children?: StudioComponentChild[];

  events?: StudioComponentEvents;
};

/**
 * This is used to track vairants for a single component
 */
export type StudioComponentVariant = {
  /**
   * The combination of vairants that comprise this variant
   */
  variantValues: { [key: string]: string };

  /**
   * The overridden properties for this variant
   */
  overrides: StudioComponentOverrides;
};

/**
 * This is a dictionary of overrides for a single parent component.
 */
export type StudioComponentOverrides = {
  /**
   * This is the reference to a full set of overrides applied to a component and its children
   * @returns A set of key value pairs representing overrides for the given child or top-level component.
   */
  [overridenComponentName: string]: { [key: string]: string };
};
