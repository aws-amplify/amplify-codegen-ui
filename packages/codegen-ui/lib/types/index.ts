/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import { StudioComponent } from './components';
import { StudioForm } from './form';
import { StudioTheme } from './theme';
import { StudioView } from './view';

export * from './actions';
export * from './components';
export * from './bindings';
export * from './events';
export * from './figma';
export * from './properties';
export * from './theme';
export * from './relational-operator';
export * from './studio-schema';
export * from './form';
export * from './view';
export * from './data';
export * from './string-format';
export * from './featureFlags';

export type StudioSchema = StudioComponent | StudioForm | StudioView | StudioTheme;
