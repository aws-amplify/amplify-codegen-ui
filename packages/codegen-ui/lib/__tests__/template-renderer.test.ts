/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import { mocked } from 'ts-jest/utils'; // eslint-disable-line import/no-extraneous-dependencies
import { existsSync, mkdirSync } from 'fs';
import { StudioComponent } from '../types';
import { StudioTemplateRendererFactory } from '../template-renderer-factory';
import { StudioTemplateRendererManager } from '../template-renderer';
import { MockOutputManager, MockTemplateRenderer } from './__utils__/mock-classes';

jest.mock('fs');

describe('StudioTemplateRendererManager', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('constructor - outputPathDir does not exist', () => {
    const outputPathDir = 'mock-output';
    const outputManager = new MockOutputManager();
    const rendererFactory = new StudioTemplateRendererFactory(
      (component: StudioComponent) => new MockTemplateRenderer(component, outputManager, {}),
    );

    mocked(existsSync).mockImplementation(() => false);
    new StudioTemplateRendererManager(rendererFactory, { outputPathDir }); // eslint-disable-line no-new
    expect(existsSync).toHaveBeenCalled();
    expect(mkdirSync).toHaveBeenCalledWith(outputPathDir);
  });

  test('constructor - outputPathDir does exist', () => {
    const outputPathDir = 'mock-output';
    const outputManager = new MockOutputManager();
    const rendererFactory = new StudioTemplateRendererFactory(
      (component: StudioComponent) => new MockTemplateRenderer(component, outputManager, {}),
    );

    mocked(existsSync).mockImplementation(() => true);
    new StudioTemplateRendererManager(rendererFactory, { outputPathDir }); // eslint-disable-line no-new
    expect(existsSync).toHaveBeenCalled();
    expect(mkdirSync).not.toHaveBeenCalled();
  });

  describe('renderSchemaToTemplate', () => {
    test('render component', () => {
      const outputPathDir = 'mock-output';
      const outputManager = new MockOutputManager();
      const rendererFactory = new StudioTemplateRendererFactory(
        (component: StudioComponent) => new MockTemplateRenderer(component, outputManager, {}),
      );
      const rendererManager = new StudioTemplateRendererManager(rendererFactory, { outputPathDir });

      const component = {
        componentType: 'Text',
        name: 'MyText',
        properties: {},
        bindingProperties: {},
      };
      const result = rendererManager.renderSchemaToTemplate(component);
      expect(result).toMatchSnapshot();
      expect(result.renderComponentToFilesystem).toHaveBeenCalledWith(outputPathDir);
    });

    test('throw error when component is not defined', () => {
      const outputPathDir = 'mock-output';
      const outputManager = new MockOutputManager();
      const rendererFactory = new StudioTemplateRendererFactory(
        (component: StudioComponent) => new MockTemplateRenderer(component, outputManager, {}),
      );
      const rendererManager = new StudioTemplateRendererManager(rendererFactory, { outputPathDir });

      expect(() => rendererManager.renderSchemaToTemplate(undefined)).toThrowErrorMatchingSnapshot();
    });
  });

  describe('renderSchemaToTemplates', () => {
    test('render components', () => {
      const outputPathDir = 'mock-output';
      const outputManager = new MockOutputManager();
      const mockRender = jest.fn(
        (component: StudioComponent) => new MockTemplateRenderer(component, outputManager, {}),
      );
      const rendererFactory = new StudioTemplateRendererFactory(mockRender);
      const rendererManager = new StudioTemplateRendererManager(rendererFactory, { outputPathDir });

      const component = {
        componentType: 'Text',
        name: 'MyText',
        properties: {},
        bindingProperties: {},
      };
      rendererManager.renderSchemaToTemplates([component]);
      expect(mockRender).toHaveBeenCalledWith(component);
    });

    test('throw error when component is not defined', () => {
      const outputPathDir = 'mock-output';
      const outputManager = new MockOutputManager();
      const rendererFactory = new StudioTemplateRendererFactory(
        (component: StudioComponent) => new MockTemplateRenderer(component, outputManager, {}),
      );
      const rendererManager = new StudioTemplateRendererManager(rendererFactory, { outputPathDir });

      expect(() => rendererManager.renderSchemaToTemplates(undefined)).toThrowErrorMatchingSnapshot();
    });
  });
});
