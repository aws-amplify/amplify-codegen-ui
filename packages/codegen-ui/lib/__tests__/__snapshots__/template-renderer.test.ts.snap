// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FrontendManagerTemplateRendererManager renderSchemaToTemplate render component 1`] = `
Object {
  "componentText": "MyText",
  "renderComponentToFilesystem": [MockFunction] {
    "calls": Array [
      Array [
        "mock-output",
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
}
`;

exports[`FrontendManagerTemplateRendererManager renderSchemaToTemplate throw error when component is not defined 1`] = `"Please ensure you have passed in a valid component schema"`;

exports[`FrontendManagerTemplateRendererManager renderSchemaToTemplates throw error when component is not defined 1`] = `"Please ensure you have passed in a valid schema"`;
